chained_filter_controller/src/chained_filter.cpp

```diff
@@ -0,0 +1,160 @@
+      {
+        rcl_interfaces::msg::ParameterDescriptor desc;
+        desc.name = "filter_chain.filter1.name";
+        desc.type = rcl_interfaces::msg::ParameterType::PARAMETER_STRING;
+        desc.read_only = true;
+        get_node()->declare_parameter(
+          "filter_chain.filter1.name", rclcpp::ParameterValue{""}, desc);
+        has_global_filter_setting =
+          !get_node()->get_parameter("filter_chain.filter1.name").as_string().empty();
+      }
+      RCLCPP_INFO_EXPRESSION(
+        get_node()->get_logger(), has_global_filter_setting,
+        "Using global filter setting for all interfaces.");
+    }
+
+    for (size_t i = 0; i < params_.input_interfaces.size(); ++i)
+    {
+      filters_[i] = std::make_unique<filters::FilterChain<double>>("double");
+      if (!filters_[i]->configure(
+            has_global_filter_setting ? "filter_chain"
+                                      : params_.input_interfaces[i] + ".filter_chain",
+            get_node()->get_node_logging_interface(), get_node()->get_node_parameters_interface()))
+      {
+        RCLCPP_ERROR(
+          get_node()->get_logger(),
+          "Failed to configure filter chain for interfaces  %s. Check the parameters for filters "
+          "setup.",
+          params_.input_interfaces[i].c_str());
+        return controller_interface::CallbackReturn::FAILURE;
+      }
+    }
+  }
+  catch (const std::exception & e)
+  {
+    fprintf(stderr, "Exception thrown during configure stage with message: %s \n", e.what());
+    return controller_interface::CallbackReturn::ERROR;
+  }
+  return controller_interface::CallbackReturn::SUCCESS;
+}
+
+controller_interface::CallbackReturn ChainedFilter::on_activate(const rclcpp_lifecycle::State &)
+{
+  std::fill(
+    output_state_values_.begin(), output_state_values_.end(),
+    std::numeric_limits<double>::quiet_NaN());
+  return controller_interface::CallbackReturn::SUCCESS;
+}
+
+controller_interface::return_type ChainedFilter::update_and_write_commands(
+  const rclcpp::Time &, const rclcpp::Duration &)
+{
+  for (size_t i = 0; i < params_.input_interfaces.size(); ++i)
+  {
+    const auto sensor_op = state_interfaces_[i].get_optional();
+    if (!sensor_op.has_value())
+    {
+      RCLCPP_ERROR(
+        get_node()->get_logger(), "Failed to read sensor value from state interface '%s'.",
+        state_interfaces_[0].get_name().c_str());
+      continue;
+    }
+    if (std::isfinite(sensor_op.value()))
+    {
+      filters_[i]->update(sensor_op.value(), output_state_values_[i]);
+    }
+  }
+
+  return controller_interface::return_type::OK;
+}
+
+std::vector<hardware_interface::StateInterface> ChainedFilter::on_export_state_interfaces()
+{
+  std::vector<hardware_interface::StateInterface> state_interfaces;
+  for (size_t i = 0; i < params_.output_interfaces.size(); ++i)
+  {
+    state_interfaces.emplace_back(
+      get_node()->get_name(), params_.output_interfaces.at(i), &output_state_values_.at(i));
```

Comment:
One curious question here 

If you remember in the past, we said only the last part of the interface needs to be used in the interface part and the rest we use in the prefix.

For instance, if we use the state interface `joint1/position` and then when we filter the exposed interface will be with prefix: `ctrl_name` and the interface as `joint1/position` instead of `ctrl_name/joint1` as prefix and `position` as interface. 

Does it make sense to follow the same strategy here?

Reply:
Here we propose the list of output_interfaces, for example we can have input_interface `joint1/position` and output interface `joint1/velocity` for a derivative filter. This would mean that we have to split the string and add only `velocity` as interface. OK for this simple example, but this might be tricky for multiple filters in a chain.

If we care (what are the implications, if we don't?), we should change the parameters to have `interface_name` (like `joint1`, it is not only limited to joints right?), and `input_interface_type` (like position) and `output_interface_type` (like velocity)?

Reply:
which string, the string given in `output_interfaces`?

Reply:
done